internal/pahlimiter/limiter_test.go

```diff
@@ -0,0 +1,329 @@
+		// as resources are acquired at least once and *usually* twice.
+		assert.Less(t, finalHistories, int64((maxConcurrentHistory*2)+1))
+		t.Log("histories:", finalHistories) // to also show successful values with -v
+	})
+}
+
+func TestWeighted(t *testing.T) {
+	t.Parallel()
+
+	t.Run("sequential works", func(t *testing.T) {
+		t.Parallel()
+
+		l, err := NewWeighted(1, 1, 10)
+		require.NoError(t, err, "should be impossible")
+		defer l.Close()
+
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+		defer cancel()
+
+		ok, release := l.Poll(ctx)
+		assert.True(t, ok)
+		release()
+
+		ok, release = l.GetHistory(ctx)
+		assert.True(t, ok)
+		release()
+	})
+	t.Run("fair thrashing should be roughly fair", func(t *testing.T) {
+		t.Parallel()
+		// this is the actual test being done: thrash on a balanced-weight limiter,
+		// and make sure the behavior is balanced.
+		//
+		// since "balanced" is hard to determine with random channel behavior,
+		// this is just run multiple times, and checked for excessive skew.
+		test := func() (int64, int64) {
+			l, err := NewWeighted(1, 1, 10)
+			require.NoError(t, err, "should be impossible")
+			defer l.Close()
+
+			return thrashForFairness(l)
+		}
+
+		// run N times in parallel to reduce noise.
+		// at least one of them must be fair, and none may have zeros.
+		var wg sync.WaitGroup
+		type result struct {
+			polls, histories int64
+		}
+		results := make([]result, 5)
+		wg.Add(len(results))
+		for i := 0; i < len(results); i++ {
+			i := i
+			go func() {
+				polls, histories := test()
+
+				// this should never happen
+				assert.NotZero(t, polls)
+				assert.NotZero(t, histories)
+
+				results[i] = result{
+					polls:     polls,
+					histories: histories,
+				}
+				wg.Done()
+			}()
+		}
+		wg.Wait()
+
+		t.Log("results:", results)
+		for _, r := range results {
+			// must be within 1/3 as much, found experimentally.
+			// out of 1000+ tries locally, only a couple have checked the third value.
+			if r.polls < (r.histories/3)*2 {
+				t.Log("low polls, checking next")
+				continue
+			}
+			if r.histories < (r.polls/3)*2 {
+				t.Log("low histories, checking next")
+				continue
+			}
+
+			t.Log("fair enough!")
+			break // no need to check the rest
+		}
+	})
+	t.Run("imbalanced thrashing should starve the higher weight", func(t *testing.T) {
+		t.Parallel()
+		// similar to fair-thrashing, but in this case we expect the lower-weight history
+		// requests to starve polls most of the time (as two need to release for a single poll to run).
+		test := func() (int64, int64) {
+			l, err := NewWeighted(2, 1, 10)
+			require.NoError(t, err, "should be impossible")
+			defer l.Close()
+
+			return thrashForFairness(l)
+		}
+
+		// run N times in parallel to reduce noise.
+		// none may have zeros, and it should be heavily skewed in favor of history requests
```

Comment:
I've used this kind of "run, measure, fail if I can't find what I expect" thing in other tests I've written (not in this repo), it has worked out fairly well.

I'm definitely open to ideas to improve it though, this was sorta a blind duplication because I knew it'd be an issue because it's inherently noisy.  There's probably some better ways.